### ISSUE TITLE
Investigate Cython + Pythran as an acceleration option

### ISSUE THREAD

**Issue by stefanv** [scikit-image/scikit-image#2956]
Currently, we heavily employ Cython for optimization, but there is now an experimental feature to support NumPy constructs via Cythran.  We should investigate.

/cc @serge-sans-paille

**Comment by serge-sans-paille** [comment]
@stefanv can you point me at some kernels that would be a good start to port?

**Comment by stefanv** [comment]
I think any existing piece of Cython code would be a good target for benchmarking vs Pythran.

**Comment by serge-sans-paille** [comment]
I've been wandering in the codebase and here is my educated guess: there is very few places where cython + pythran integration would be of any help, as most of the cython kernels don't make use of ufunc and cannot be transformed into using it with ease. yet most of the kernels are well withing the bounds of what pythran can compile, I provided a sample impelemntation here https://github.com/serge-sans-paille/scikit-image/commit/793612df141bb722a3d50fa83fbb3225c594f8f4 .

Let me know if you want me to dig that way!

**Comment by jni** [comment]
@serge-sans-paille that's very clean! Nice!

However, I'm unable to compile, getting the error `NameError: name 'intp' is not defined`. Full traceback below:

```pytb
 ~/projects/scikit-image (feature/test-pythran-support)
 $ pip install --no-deps -U -e .
Obtaining file:///home/jni/projects/scikit-image
    Complete output from command python setup.py egg_info:
    /home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/setuptools/dist.py:388: UserWarning: Normalizing '0.15dev' to '0.15.dev0'
      normalized_version,
    Partial import of skimage during the build process.
    WARNING: Disabling color, you really want to install colorlog.
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/home/jni/projects/scikit-image/setup.py", line 161, in <module>
        **extra
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/numpy/distutils/core.py", line 135, in setup
        config = configuration()
      File "/home/jni/projects/scikit-image/setup.py", line 81, in configuration
        config.add_subpackage('skimage')
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/numpy/distutils/misc_util.py", line 1024, in add_subpackage
        caller_level = 2)
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/numpy/distutils/misc_util.py", line 993, in get_subpackage
        caller_level = caller_level + 1)
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/numpy/distutils/misc_util.py", line 930, in _get_configuration_from_setup_py
        config = setup_module.configuration(*args)
      File "skimage/setup.py", line 14, in configuration
        config.add_subpackage('feature')
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/numpy/distutils/misc_util.py", line 1024, in add_subpackage
        caller_level = 2)
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/numpy/distutils/misc_util.py", line 993, in get_subpackage
        caller_level = caller_level + 1)
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/numpy/distutils/misc_util.py", line 930, in _get_configuration_from_setup_py
        config = setup_module.configuration(*args)
      File "skimage/feature/setup.py", line 30, in configuration
        config.ext_modules.append(PythranExtension('_brief_cy', ['skimage/feature/brief_cy.py']))
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/pythran/dist.py", line 58, in __init__
        module_name, cpponly=True)
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/pythran/toolchain.py", line 464, in compile_pythranfile
        **kwargs)
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/pythran/toolchain.py", line 393, in compile_pythrancode
        specs = spec_parser(pythrancode)
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/pythran/spec.py", line 460, in spec_parser
        return SpecParser()(path_or_text)
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/pythran/spec.py", line 417, in __call__
        self.parser.parse(pythran_data, lexer=self.lexer, debug=False)
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/ply/yacc.py", line 333, in parse
        return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/ply/yacc.py", line 1120, in parseopt_notrack
        p.callable(pslice)
      File "/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/pythran/spec.py", line 277, in p_dtype
        p[0] = eval(p[1]),
      File "<string>", line 1, in <module>
    NameError: name 'intp' is not defined
```

any ideas?

**Comment by serge-sans-paille** [comment]
what version of pythran are tou using?

**Comment by jni** [comment]
What was on conda-forge: 0.8.5.

**Comment by jni** [comment]
Cool! Well, it seems a lot cleaner than comparable Cython, and for this one benchmark at least, performance is identical:

```python
In [1]: from skimage.feature import BRIEF
In [2]: import numpy as np
In [3]: keypoints = np.random.randint(10, 502, size=(1000, 2))
In [4]: from skimage import data
In [5]: astro = data.astronaut()[:, :, 1]
In [6]: brief = BRIEF()
In [7]: %timeit brief.extract(astro, keypoints)
6.7 ms ± 24.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

For Pythran 0.8.5 (with intp replaced by int), the output is:

```
6.6 ms ± 17.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by jni** [comment]
Oh, question: what's the deal with Pythran and the GIL?

**Comment by serge-sans-paille** [comment]
Pythran releases the GIL once it has done the conversion of the arguments, and takes it back when it needs to convert the return value.

http://pythran.readthedocs.io/en/latest/MANUAL.html#gil-interaction

**Comment by jni** [comment]
Wow. :blush:

**Comment by jni** [comment]
@serge-sans-paille we won't stop you from doing work for us. =P

In all seriousness, I think the deeper question is whether we would be willing to take on Pythran as a new build dependency. To me, on first pass, it seems doable, since it does not require any new runtime dependencies. The same questions that we have/had for Numba apply:

- what is the state of debugging for Pythran code?
- what about profiling? Is it easy to diagnose slowdowns in the same way that cython --annotate helps to do? Is it easy to profile the generated C code?
- does it support things like fused types? (compiling the same function for float32 and float64 input, for example.)
- what are the advantages over Cython? (Sorry if this is covered in your paper — I haven't had a chance to read it yet.)

From a development perspective, it does look easier, which is super nice. For example, from your example, it seems that Pythran code is also Python code, so people could run skimage without doing any compilation? That's huge. (Although we would want to be able to raise a warning in situations where three has not been a compilation.)

**Comment by serge-sans-paille** [comment]
- debugging : it's native code, so you ``just'' need to run ``gdb --arg python my_python_script_that_calls_a_native_lib.py`` and set your breakpoint etc.

- profiling : there is no such tool as ``cython -a``, which makes sense because everything is pure C++, no Python C-API. So the basic tool like ``callgrind`` apply. There is still a ``pythran -P`` command that lets you observe the transformed file *before* it gets dump to C++ and ``pythran -E`` that generates the .cpp file for inspection, tracing etc.

- fused types are supported, in the sense that

  ```python
  #pythran export foo(int[])
  #pythran export foo(float[])
  def foo(x): return x
   ```
   works as expected.

- advantages over cython: cleanr syntax, and thiner binaries (see http://serge-sans-paille.github.io/pythran-stories/shrinking-pythran-generated-binaries.html)

- disadvtanges over cython : longer compilation time, no mixed mode (if a function / package is not supported, you need to implement it), no support for user classes

> people could run skimage without doing any compilation

Yeah, that should work. needs some magic in the setup file, like checking if pythran is available, and if not just don't generate extensions. But given the look of scikit-image kernels, it would be ultra slow...

**Comment by jni** [comment]
> there is no such tool as cython -a, which makes sense because everything is pure C++, no Python C-API.

Ah, so rather than producing slow code, Pythran will just fail to compile if there is any Python interaction?

> fused types are supported

That's great! Clean interface, too.

> given the look of scikit-image kernels, it would be ultra slow...

Yes, this is not to be used in production, but rather to measure test coverage (on tiny examples). When run in this way we would raise a big fat warning linking to the appropriate documentation. This would be a massive improvement over the current behaviour, which is an ImportError!

**Comment by serge-sans-paille** [comment]
> Ah, so rather than producing slow code, Pythran will just fail to compile if there is any Python interaction?

Yes

**Comment by serge-sans-paille** [comment]
Pushed a few more kernels

**Comment by serge-sans-paille** [comment]
@jni / @stefanv what could be the next step? Me porting more kernels ?

**Comment by jni** [comment]
@serge-sans-paille, @stefanv and I are very excited about this! However taking on a new build dependency is a Big Deal and we need to have a discussion about this on the mailing list. My guess is that nothing should happen until the next core-dev meeting (this is something that we are about to announce), and possibly until SciPy 2018 (Jul 9-15). 

My proposal for a course of action is as follows:
- do as many kernels as you feel like, though for now these are great, and there's no need for more. We don't need to port all of scikit-image at once.
- wait for https://github.com/scikit-image/scikit-image/pull/3137 to get merged (should happen very soon)
- open a PR to skimage with your kernel changes
- include some benchmarks for each new kernel, in the `benchmarks/` directory. There will be a template there to show how to write one, and if you need help just open the PR and we can help you there. The gallery examples are a good source of inspiration for this.

At this point we can have a discussion about adding it. I'm :+1: on it. @stefanv made a good point in a recent chat about this: the fact that Pythran works with pure Python is a massive advantage, because it is means zero lock-in: any new compilers can just pick up the code and run with it.

**Comment by jni** [comment]
As a separate discussion point: two frequent requirements in image algorithms are a queue and a priority queue. Is there any easy way to implement these in Pythran?

**Comment by serge-sans-paille** [comment]
> queue and a priority queue

if this means adding support for the ``queue`` and ``heapq`` modules, then yes, that's possible.

**Comment by serge-sans-paille** [comment]
@stefanv and I are very excited about this!

So am I :-)

> do as many kernels as you feel like, though for now these are great, and there's no need for more. We don't need to port all of scikit-image at once.

I'll go on that way because it provides a lot of test case for pythran, so whatever the result, it's good for the project :-)

I'm fine with the roadmap, will create the PR once #3173 is merged and keep the ball rolling.

**Comment by serge-sans-paille** [comment]
#3226 opened!

**Comment by hmaarrfk** [comment]
@serge-sans-paille just listening to your talk from 2015 in Paris. Pretty awesome stuff.

I hope your PR goes through. I never liked having to write low level code in Cython again. I hope your PR goes through :D Excited to use it as a template.

**Comment by grlee77** [comment]
closed by #3226
